text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,384 @@
+    pub fn remove_event_listener(this: &EventTarget, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = dispatchEvent)]
+    pub fn dispatch_event(this: &EventTarget, event: Event);
+}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    type Node;
+
+
+    // Methods from EventTarget
+    #[wasm_bindgen(method, js_name = addEventListener)]
+    pub fn add_event_listener(this: &Node, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = removeEventListener)]
+    pub fn remove_event_listener(this: &Node, type: &str, callback: &Closure<FnMut(Event)>, options: bool);
+
+    #[wasm_bindgen(method, js_name = dispatchEvent)]
+    pub fn dispatch_event(this: &Node, event: Event);
+
+
+    // Methods from Node
+    #[wasm_bindgen(method, getter = nodeName)]
+    pub fn node_value(this: &Node) -> JsString;
+
+    #[wasm_bindgen(method, getter = textContent)]
+    pub fn text_content(this: &Node) -> JsString;
+
+    #[wasm_bindgen(method, js_name = appendChild)]
+    pub fn append_child(this: &Node, node: Node) -> Node;
+
+    #[wasm_bindgen(method, js_name = removeChild)]
+    pub fn remove_child(this: &Node, child: Node) -> Node;
+}
+```
+
+As you can see, it duplicates the `add_event_listener`, `remove_event_listener`, and `dispatch_event`
+methods on `Node`.
+
+Similarly, it would have to duplicate all of the `EventTarget` and `Node` methods on `Element`. And it
+would have to duplicate all of the `EventTarget`, `Node`, and `Element` methods on `HTMLElement`, etc.
+
+This is an incredible amount of duplication, so it's only really feasible for a tool which automatically
+generates the methods (such as the WebIDL generator). Trying to do this duplication by hand is unmaintainable.
+
+That means that if you're using inherit impls, it will be very painful to use method inheritance with anything
+other than WebIDL, because of the maintenance burden.
+
+And because class-based inheritance is used outside of WebIDL, we want to be able to support non-WebIDL use
+cases.
+
+# Unresolved Questions
+[unresolved]: #unresolved-questions
+
+In the above examples I used the `I` prefix for the traits (e.g. `IHtmlElement`, `IElement`, etc.)
+
+This is because JavaScript ties classes and methods together, but Rust keeps them separate.
+
+As an example, in Rust we have an `HTMLElement` type (which corresponds to `HTMLElement` in JavaScript),
+but methods are split into a separate trait, and we cannot re-use the `HTMLElement` name for the
+trait, so we essentially need two namespaces: one for types and one for traits.
+
+And so, as a convention, I added the `I` prefix to the traits, which essentially put them into a
+separate namespace from the types.
+
+There are other conventions, and I'm not sure about the best way to solve this problem.
+
+----
+
+This proposal requires a lot of boilerplate (to define the inherit and trait methods). This is acceptable
```

Comment:
Maybe move this stuff out of unresolved questions and into a new `Future Extensions` section?